scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,97 @@
+# 0001 - ScannerV4 APIs definition
+
+- **Status:** Accepted
+- **Created:** [2023-05-09 Tue]
+
+## Context
+
+We are transitioning StackRox Scanner to use [ClairCore](https://github.com/quay/claircore) as its underlying scanning engine. The goal is to align vulnerability scanning results between Quay/Clair and StackRox. The transition will be gradual, starting with container image scanning. During this period, the existing Scanner (ScannerV2) and the new ClairCore-based Scanner (ScannerV4) will be simultaneously deployed.
+
+The move to ClairCore and the adoption of ScannerV4 bring the need for new API definitions. Also, ClairCore allows for Clair deployment modes that better support the secured cluster use-case (local scanning), horizontal scalability, and high-availability scenarios. The new APIs ideally should align with ClairCore's capabilities while minimizing changes in Central.
+
+## Decision
+
+ScannerV4 APIs will exclusively use gRPC. ScannerV4 APIs are not backward compatible with ScannerV2.
+
+ScannerV4 will offer modes of operation [akin to Clair's deployment models](https://quay.github.io/clair/howto/deployment.html). Each mode will implement different gRPC services, reflecting the underlying separation between ClairCore's [libindex](https://pkg.go.dev/github.com/quay/claircore/libindex#Libindex) and [libvuln](https://pkg.go.dev/github.com/quay/claircore/libvuln#Libvuln). They will be named "indexer" and "matcher". Both modes can be enabled concurrently.
+
+ScannerV4 will use "index reports" and "vulnerability reports" data models, similar to ClairCore's. Both types will link to a "scannable resource" (e.g., a container image) using an ID created by Scanner clients (i.e. `hash_id`). This is a string that uniquely identifies the resource's manifest. To avoid conflicts between different resources, and allow changes to how the IDs are generated, Central and Sensor (the existing clients) will namespace them with `/v1/<resource>/`.  Initially, `/v1/containerimage/<image-digest>` will be supported for container images.
+
+Index Reports are persisted, and their lifecycle is managed by Scanner. Central and Sensor create reports on demand. Scanner is responsible for deleting least-recently used reports.
+
+ScannerV4 gRPC endpoints will use gRPC status codes to communicate the return status in case of non-successful responses.
+
+ScannerV4 will also provide endpoints for health and observability checks, which are not detailed in this ADR.
+
+In Central, ScannerV4 will replace ScannerV2 as the container image scanner. A new image integration for ScannerV4 will be created and replace ScannerV2's integration (type "clarify"), activated by the feature flag. The "clarify" type will continue to be used for orchestrator and node scanning.
```

Comment:
nit: "the feature flag" -> "a feature flag"

Reply:
b47cc004a9..77252800a7